### PR TITLE
telemetry: use an extra cache directory for telemetry

### LINKED ISSUE #8396: Harden the SDK against a read-only DAML_PATH
This has been brought up because certain installations will have ~/.daml (or the equivalent DAML_PATH) as read-only. Ideally the SDK should still work, although slightly hampered by an inability to cache or update, but right now there appears to be several problems, such as telemetry blocking `daml damlc ide` because it can't write a `.machine_id` file.

My suggestion is to try to capture some of these issues by writing an integration test that turns the DAML_PATH directory to be read-only (after installation) and running some tests with daml damlc ide, daml assistant, etc.

### PR DESCRIPTION
**Pull request by robin-da**
If the sdk's `.daml` directory is not writable, telemetry fails and
kills the IDE. We add a new assistant environment variable "DAML_CACHE".
If set, this directory is used for telemetry cache data.

Fixes #8396.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-assistant/exe/DA/Daml/Assistant.hs
Comment (sofiafaro-da) [inline daml-assistant/exe/DA/Daml/Assistant.hs:305]: This seems incorrect -- you're setting GCP cache path to the DAML home.

File: libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs
Comment (sofiafaro-da) [inline libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs:403]: How is writability ensured?

**Review by sofiafaro-da — COMMENTED** [review]
I'm not sure this resolves the original problem. What happens when DAML_CACHE isn't set, or is set to a read-only directory? Obviously that's a bad situation but the point of the ticket is that this *must not* break the IDE.

**Comment by cocreature** [comment]
I have a conceptual question here: I would expect that setups where you cannot write to the installation directory are also very unlikely to want to enable telemetry. So rather than trying to come up with a way to find a writable directory for this, can we jus thandle the case where it is not writable and disable telemetry in this case?

**Comment by robin-da** [comment]
The idea here is to enable the SDK to work in restricted environments with the introduction of the DAML_CACHE environment variable. But I'm also in favour of turning telemetry off completely when we encounter a read-only cache directory.

File: daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs
Comment (cocreature) [inline daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs:48]: If we want a chance of this working automatically, I think we should default this to `XDG_CONFIG_HOME` or `XDG_CACHE_HOME`. Then an enterprise setup could customize `DAML_HOME` to ensure binaries are not user writable while `DAML_CACHE` will most likely not need customization.

The one thing I’m not entirely sure about is backwards compatibility but tbh, I think we can probably just ignore that given that we only write telemetry files there atm and writing those again is perfectly fine.

File: daml-assistant/exe/DA/Daml/Assistant.hs
Comment (cocreature) [reply daml-assistant/exe/DA/Daml/Assistant.hs:305]: see my comment above, I think we should choose the default differently.

File: libs-haskell/da-hs-base/tests/Telemetry.hs
Comment (cocreature) [inline libs-haskell/da-hs-base/tests/Telemetry.hs:74]: maybe set it to a temp directory instead?

**Review by cocreature — COMMENTED** [review]
Having thought about this for a bit, I think this is a sensible change that could improve the situation and could be worth including even if we handle the directory not being writable. I do think we should change the default however, to get a bit more of a chance of this working automatically.

**Comment by robin-da** [comment]
@cocreature Note, that this PR now also includes the turn-off of telemetry if we end up with a read-only cache dir.

File: daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs
Comment (robin-da) [reply daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs:48]: agreed!

File: libs-haskell/da-hs-base/tests/Telemetry.hs
Comment (robin-da) [reply libs-haskell/da-hs-base/tests/Telemetry.hs]: fixed.

File: daml-assistant/exe/DA/Daml/Assistant.hs
Comment (robin-da) [reply daml-assistant/exe/DA/Daml/Assistant.hs]: set to `Nothing`, defaults to XDG_CACHE_DIR now.

File: libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs
Comment (robin-da) [reply libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs]: It checks the permissions now and returns `Nothing` if we end up with a read-only cache.

File: daml-assistant/exe/DA/Daml/Assistant.hs
Comment (cocreature) [inline daml-assistant/exe/DA/Daml/Assistant.hs:304]: this seems off, we don’t set the cache path at all and we use `damlPath` instead. Telemetry in the assistant shouldn’t behave different from telemetry in damlc.

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTestUtils.hs
Comment (cocreature) [inline daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTestUtils.hs:44]: shouldn’t do any harm but we don’t test telemetry in the integration tests so this is not testing the code path we care about.

File: libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs
Comment (cocreature) [inline libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs:415]: I’m slightly worried about the fact that we check this once here for the directory but don’t handle failures on the individual file writes and reads. That probably works in reasonable setups but I wouldn’t be surprised if we end up in situations where the directory is writable but individual files are not writable or not readable. Did you consider to instead guard the actual file operations and handle failures there? Afaict, we only have 4 kinds of file access:

1. `setOptIn`: If we cannot write `.opted_in` or remove `.opted_out` just do nothing.
2. `logOptOut`: If we cannot read the gcp metadata do nothing, if we cannot write the opted out file (we currently seem to do that asynchronously, that should probably change), do nothing.
3. `logMetadata`: If we cannot read the gcp metadata do nothing.
4. `sendData`: If we cannot access the sent data file, do nothing.

We can still keep the check here, if the directory is not writable there is no point in turning telemetry on but this should be a bit more robust in weird setups.

File: libs-haskell/da-hs-base/tests/Telemetry.hs
Comment (cocreature) [inline libs-haskell/da-hs-base/tests/Telemetry.hs:56]: A test that checks that a non-writable cache directory is supported as well would be great.

**Review by cocreature — COMMENTED** [review]
LGTM, one comment on whether we can and maybe should make it even safer and I’d like to have a test for a non-writable cache directory.

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTestUtils.hs
Comment (robin-da) [reply daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTestUtils.hs:44]: yes, it's meant to detect future dependency on a writable .daml directory.

File: daml-assistant/exe/DA/Daml/Assistant.hs
Comment (robin-da) [reply daml-assistant/exe/DA/Daml/Assistant.hs:304]: By setting it to `Nothing`, we will try to use the XDG_CACHE_DIR and use this when set.

File: libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs
Comment (robin-da) [reply libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs:415]: Yes, that was my first approach but it became quite a mess. The idea is that it's our code generating these files. If the directory is not writable in the first place, we won't generate the files either. If we can create them, then they will be writable. If someone changes the permissions of these files afterwards, I think failing is acceptable.

File: libs-haskell/da-hs-base/tests/Telemetry.hs
Comment (robin-da) [reply libs-haskell/da-hs-base/tests/Telemetry.hs:56]: I'll add one.

File: libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs
Comment (cocreature) [reply libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs:415]: Alright, that seems convincing. Do we have to create this directory? I know that previously we didn’t have to create it because the assistant created it but now that it’s a separate directory, I’m not entirely clear if we can rely on it existing. This check probably then also needs to take account the directory not existing/run after we’ve created it.

File: daml-assistant/exe/DA/Daml/Assistant.hs
Comment (cocreature) [reply daml-assistant/exe/DA/Daml/Assistant.hs:304]: Ah right but the `optedOutPath` and `optedInPath` below still need to take the actual cache directory into account instead of being relative to `damlPath`.

Comment (robin-da) [reply daml-assistant/exe/DA/Daml/Assistant.hs:304]: ah, that's true. Good catch!

File: libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs
Comment (robin-da) [reply libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs:420]: We create the directory if it doesn't already exist in `getCacheDir` in the gcp logger.

File: daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs
Comment (sofiafaro-da) [inline daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs:50]: I disagree with using `XDG_CACHE_HOME` here directly, and it will cause problems since it's not an environment variable "owned" by the assistant.

This should really be `DAML_CACHE`, and the assistant should set up the `DAML_CACHE` environment variable when it runs (using `XDG_CACHE_HOME` as default), the way it already sets up e.g. `DAML_PROJECT` and `DAML_SDK_VERSION` and so on.

```suggestion
damlCacheEnvVar = "DAML_CACHE"
```

File: daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs
Comment (sofiafaro-da) [inline daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs:106]: This is not going to work! The assistant filters out damlEnvVars that it doesn't set. This list should only contain variables that are "owned" by the assistant. See my comment above.

**Review by sofiafaro-da — CHANGES_REQUESTED** [review]
We can't have XDG_CACHE_HOME be considered a "DAML environment variable." Having a DAML_CACHE env var makes sense, where it gets set up by the assistant, with XDG_CACHE_HOME as the default.

A couple more things:

- We should let people know (in the changelog) that they may need to opt-in/opt-out of telemetry again.

- Once we have `DAML_CACHE`, and in a separate PR, we should change the assistant to use DAML_CACHE instead of `DAML_HOME/cache`, to store its version.json / snapshot.json cache.

File: daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs
Comment (cocreature) [reply daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs:50]: good catch, this should definitely be set by the assistant.

**Comment by cocreature** [comment]
>     * We should let people know (in the changelog) that they may need to opt-in/opt-out of telemetry again.

In most cases this shouldn’t be necessary. We store the telemetry setting in the vscode extension which then in turn sets the `--telemetry` option passed to damlc. So if you enabled telemetry in vscode, it should start damlc with the option which will create the file on startup without prompting you again. The one case where you have to enable it again is if you only use the assistant but not daml studio but not sure that deserves a mention in the release notes.

File: daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs
Comment (robin-da) [reply daml-assistant/daml-project-config/DA/Daml/Project/Consts.hs:50]: ok, let's do it this way then.

**Comment by robin-da** [comment]
@sofiafaro-da 
```After we have DAML_CACHE, in a separate PR, we should change the assistant to use DAML_CACHE instead of DAML_HOME/cache, to store its versions.json / snapshots.json cache.```

That's a good idea, I'll take care of it.

File: daml-assistant/src/DA/Daml/Assistant/Env.hs
Comment (cocreature) [inline daml-assistant/src/DA/Daml/Assistant/Env.hs:149]:
```suggestion
-- | Get the DAML cache folder. This defaults to $XDG_CACHE_HOME/daml.
```

File: libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs
Comment (cocreature) [inline libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs:193]: Afaict, this check is going to fail on an empty setup. You enable telemetry in DAML Studio, we run `damlc ide --telemetry=yes`, we get here, the files are not created and we don’t enable telemetry.

We need to keep the logic to write the file if we are opted in.

File: compiler/damlc/lib/DA/Cli/Damlc.hs
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc.hs:450]: We don’t actually use the env var here do we? I would have expected this to read `DAML_CACHE` and maybe default to this or `DAML_HOME` if it’s not set.

Comment (robin-da) [reply compiler/damlc/lib/DA/Cli/Damlc.hs:450]: I think you're right, `DAML_CACHE` makes more sense.

File: libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs
Comment (robin-da) [reply libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs]: I think you're right. I moved the logic back to the assistant.

File: compiler/damlc/lib/DA/Cli/Damlc.hs
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc.hs:450]: can we use the environment variable constants we defined in daml-project-config?

File: compiler/damlc/lib/DA/Cli/Damlc.hs
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc.hs:452]: similar question here, can we deduplicate this with the assistant?

File: compiler/damlc/lib/DA/Cli/Damlc.hs
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc.hs:458]: Afaict, we only need to set $HOME everywhere because `getXdgDirectory` needs it. Can we just disable telemetry if `getXdgDirectory` fails instead? I’m slightly worried users might hit this in CI as well.

File: daml-assistant/daml-project-config/DA/Daml/Project/Types.hs
Comment (cocreature) [inline daml-assistant/daml-project-config/DA/Daml/Project/Types.hs:91]:
```suggestion
-- | File path to a cache directory, e.g. ~/.cache/daml.
```

**Review by cocreature — APPROVED** [review]
Very nice, thanks for pushing this through! I’d like to get rid of the `$HOME` requirement before merging but should be good apart from that.

File: compiler/damlc/lib/DA/Cli/Damlc.hs
Comment (robin-da) [reply compiler/damlc/lib/DA/Cli/Damlc.hs:450]: changed.

File: compiler/damlc/lib/DA/Cli/Damlc.hs
Comment (robin-da) [reply compiler/damlc/lib/DA/Cli/Damlc.hs:455]: changed, I got really annoyed of setting $HOME everywhere ...

File: compiler/damlc/lib/DA/Cli/Damlc.hs
Comment (robin-da) [reply compiler/damlc/lib/DA/Cli/Damlc.hs]: now it's only getting DAML_CACHE.

File: compiler/damlc/lib/DA/Cli/Damlc.hs
Comment (cocreature) [reply compiler/damlc/lib/DA/Cli/Damlc.hs:455]: So this means running damlc without having `DAML_HOME` set won’t enable telemetry? That seems perfectly acceptable to me :+1:

File: daml-assistant/src/DA/Daml/Assistant/Env.hs
Comment (cocreature) [inline daml-assistant/src/DA/Daml/Assistant/Env.hs:155]:
```suggestion
            errOrCacheDir <- tryIO $ do
```

File: daml-assistant/src/DA/Daml/Assistant/Env.hs
Comment (cocreature) [inline daml-assistant/src/DA/Daml/Assistant/Env.hs:158]: looking through the code I’m actuall a bit confusey. Which part is failing if `$HOME` is not set? Afaict, both `getAppUserDirectory` and `getXdgDirectory` end up calling [getHomeDirectoryInternal](https://hackage.haskell.org/package/directory-1.3.6.1/docs/src/System.Directory.Internal.Posix.html#getHomeDirectoryInternal) which handles HOME not being set the same way.

Comment (cocreature) [reply daml-assistant/src/DA/Daml/Assistant/Env.hs:158]: ah https://github.com/haskell/directory/issues/102, I think you can upgrade directory and sidestep this whole mess.

Comment (robin-da) [reply daml-assistant/src/DA/Daml/Assistant/Env.hs:157]: yes please!

File: stack-snapshot.yaml
Comment (cocreature) [inline stack-snapshot.yaml:15]: you have to repin on both linux and windows for this to have any effect

Comment (robin-da) [reply stack-snapshot.yaml:15]: thanks!

Comment (robin-da) [reply stack-snapshot.yaml:15]: thanks!

**Review by sofiafaro-da — APPROVED** [review]
Looks great, thank you!

**Comment by robin-da** [comment]
This still fails when $HOME is not set, also for the sdk head.

**Comment by cocreature** [comment]
sdk head as in `daml-sdk-head`, i..e, it’s not related to the compatibility tests?

**Comment by robin-da** [comment]
yes, this seems unrelated to the compatibility tests.

**Comment by cocreature** [comment]
In that case, defaulting to `DAML_HOME` seems sensible as a default. I think I understand why this doesn’t fall into the trap: We locate `DAML_HOME` relative to the `daml` executable first before we call `getAppUserDirectory`.

**Comment by robin-da** [comment]
But how does `getXdgDirectory` fail?

**Comment by cocreature** [comment]
Ahrg, I think I got it:
```
moritz@adjunction ~/daml (read_only_cache)> bazel query --output build @stackage//:directory
# /home/moritz/.cache/bazel/_bazel_moritz/bb4e4404f889dc1b816f246b08c0d9ea/external/stackage/BUILD.bazel:1120:26
haskell_toolchain_library(
  name = "directory",
  visibility = ["//visibility:public"],
)
# Rule haskell_toolchain_library defined at (most recent call last):
#   /home/moritz/.cache/bazel/_bazel_moritz/bb4e4404f889dc1b816f246b08c0d9ea/external/rules_haskell/haskell/defs.bzl:609:33 in <toplevel>
```

So changing the directory version doesn’t actually work since `rules_haskell` appears to have a hardcoded list of core libraries https://github.com/tweag/rules_haskell/blob/fa8d2c53449ee1659f5f503cfe3141d27cb8b869/haskell/cabal.bzl#L847 and directory is one of them. Those core packages end up being treated differently and come from the GHC installation.

@aherrmann-da Is there a way to upgrade this without upgrading GHC or are we stuck on the older directory version for now? I know that in general you can upgrade directory just fine as long as you don’t depend on ghc (and we only depend on ghc-lib) but iirc nix also has some issues around that.

**Comment by cocreature** [comment]
I just realized that there is one more thing we should include here: When creating the cache directory, copy `.machine_id` from the old location if it exists (handle read failures by doing nothing). Without this, our user stats will be messed up by this.

**Comment by aherrmann** [comment]
> @aherrmann-da Is there a way to upgrade this without upgrading GHC or are we stuck on the older directory version for now? I know that in general you can upgrade directory just fine as long as you don’t depend on ghc (and we only depend on ghc-lib) but iirc nix also has some issues around that.

I think `vendored_packages` can be used for this. I tried with a dummy `directory` package on `rules_haskell` and at least the generated `@stackage//:BUILD.bazel` refers to the vendored package rather than the toolchain library. I think we could also make this generally more flexible and use the output of `stack ls dependencies json` to determine which packages are core instead of a hard-coded list. If stack allows overriding core packages than this should also allow that use-case in rules_haskell.

**Comment by cocreature** [comment]
> > 
> 
> I think `vendored_packages` can be used for this. I tried with a dummy `directory` package on `rules_haskell` and at least the generated `@stackage//:BUILD.bazel` refers to the vendored package rather than the toolchain library. I think we could also make this generally more flexible and use the output of `stack ls dependencies json` to determine which packages are core instead of a hard-coded list. If stack allows overriding core packages than this should also allow that use-case in rules_haskell.

Thanks! @robin-da I’ll leave it up to you if you want to try this in this PR or whether we just catch the exception and fall back to `DAML_HOME`.

**Comment by robin-da** [comment]
@cocreature I see if I can make it work with `vendored_packages`. I'll also have a look at the machine_id issue.

File: libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs
Comment (cocreature) [inline libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs:177]: This should also have a check that `newMachineIdPath` does not yet exist. We don’t want to overwrite an already existing file. It’s probably rarely an issue but better be safe than sorry.

**Review by cocreature — COMMENTED** [review]
Mostly lgtm, two coments:

1. I don’t understand why we have to set `DAML_CACHE` everywhere. Do our users now also need this in their CI envs? I was expecting that this just works with the fallback.
2. I’d recommend running this once locally to verify that `.machine_id` gets copied over correctly. Code looks fine but it’s hard to test this.

File: libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs
Comment (robin-da) [reply libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs:178]: I added the check, that's a good idea. I'll also test locally before merging.

**Comment by robin-da** [comment]
I'm still trying to figure out why the compatibility tests fail without a set DAML_CACHE dir.

File: libs-haskell/da-hs-base/tests/Telemetry.hs
Comment (cocreature) [inline libs-haskell/da-hs-base/tests/Telemetry.hs:63]: Maybe a bit clearer to set `DAML_CACHE` to a read-only directory instead for the test?

**Review by cocreature — APPROVED** [review]
Awesome, thank you so much for pushing this through!

**Comment by robin-da** [comment]
machin_id file moving has been tested locally.

File: libs-haskell/da-hs-base/tests/Telemetry.hs
Comment (robin-da) [reply libs-haskell/da-hs-base/tests/Telemetry.hs:63]: this test doesn't check for environment variables (`fakeHome` is used in the already existing test). I set the DAML_CACHE directory in the assistant tests to read-only instead.

Comment (cocreature) [reply libs-haskell/da-hs-base/tests/Telemetry.hs:63]: Ah good point, I missed the fact that we pass it explicitly in the read-only test here :+1:

**Pull request merged**
